Add log upload support for build target failed

erman-gurses opened the pull request:
This PR adds the log upload support when the build target failed within the workflow. It is not using `teatime.py` script

erman-gurses commented:
The question is how would I test this? Maybe I can purposely break purposely in the build and see how it behaves.

erman-gurses commented:
Initial test is passed.
https://github.com/ROCm/TheRock/actions/runs/14939987951/job/41975427445

ScottTodd commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -136,14 +142,30 @@ jobs:
 
       # TODO: Move to script
       - name: Upload Logs
+        if: always()
```
It's time to move this logic to a Python script.

erman-gurses replied:
so you want me write this in Python as a separated file AFAIU. Where should I locate it?

ScottTodd replied:
Pick somewhere under https://github.com/ROCm/TheRock/tree/main/build_tools. Maybe `upload_logs_to_s3.py`

ScottTodd commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -136,14 +142,30 @@ jobs:
-            --include "*.log"
-          aws s3 cp build/logs/index.html s3://therock-artifacts/${{github.run_id}}/logs/${{env.AMDGPU_FAMILIES}}/
-
+          if [ ! -d build/logs ]; then
+            echo "[INFO] Logs directory missing. Creating build/logs..."
+            mkdir -p build/logs
+          fi
```
Why create an empty directory and then upload it?

ScottTodd replied:
If there aren't any files to upload, don't upload any files?

Generally, think through the process here and what the state should be at each step:

1. Job starts. Do we set up some metadata here and create a folder in S3 with that metadata now? Did a prior job already do that and are we just writing into an already established location?
2. Job installs requirements, initializes caches, etc.
3. Job starts building. Build logs start to get produced. Build artifacts start to get produced.
4. Job finishes building. Artifacts and logs are uploaded.

What should happen if the build fails? Should we upload partial artifacts? What about logs?
What should happen if the job is cancelled? Should we upload partial artifacts? What about logs?

Notice that with log (and artifact) streaming, the answers change.

erman-gurses replied:
>What should happen if the build fails? Should we upload partial artifacts? What about logs?
>What should happen if the job is cancelled? Should we upload partial artifacts? What about logs?

1. So if the build fails in the beginning, there would not be any log to upload -  I assume the time even the  `build/logs` does not exist
2. If the build fails in the middle, there must be some log to upload so  `build/logs` exists and upload whatever it exist for logs.
3. If the job is canceled, in the middle of the building - we still should upload what we have as log.

erman-gurses replied:
I think `always()` does not work for the user triggered cancelations.

erman-gurses replied:
Tested the cancel case - worked fine and uploaded the logs: https://github.com/ROCm/TheRock/actions/runs/14985473348/job/42098511841

ScottTodd commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -136,14 +142,30 @@ jobs:
+
+          echo "[INFO] Uploading logs..."
+
+          if compgen -G "build/logs/*.log" > /dev/null; then
```
This `compgen` command is not portable to Windows. Switch to Python.

erman-gurses replied:
Sounds good!

erman-gurses commented:
Tested the normal working case:
https://github.com/ROCm/TheRock/actions/runs/14986147067/job/42100345141

marbre commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,46 @@
+def upload_logs_to_s3(bucket_name, run_id, amdgpu_family, log_dir="build/logs"):
+    s3_base_path = f"s3://{bucket_name}/{run_id}/logs/{amdgpu_family}"
+
+    if not os.path.isdir(log_dir):
+        print("[INFO] No logs directory found. Creating it.")
+        os.makedirs(log_dir)
+        return
```
Form the previous discussion, I don't see a case why you want to create an empty log dir. If the directory isn't existing there will be nothing in it and thus there is nothing to upload.

marbre commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -136,14 +143,13 @@ jobs:
+          GITHUB_RUN_ID: ${{ github.run_id }}
+          AMDGPU_FAMILIES: ${{ env.AMDGPU_FAMILIES }}
+          S3_BUCKET: therock-artifacts
       - name: Add Links to Job Summary
```
I think this step is partially needed. If you upload logs even though the build fails, you also want to add a link to the build logs (while skip adding a link to the artifacts).

marbre replied:
I think that should work.

marbre commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,46 @@
+if __name__ == "__main__":
+    bucket = os.getenv("S3_BUCKET", "therock-artifacts")
+    run_id = os.getenv("GITHUB_RUN_ID")
+    amdgpu_family = os.getenv("AMDGPU_FAMILIES", "unknown")
```
Should we rather let it fail instead of pushing to a default location? I am undecided here if it makes sense to upload logs in that case at all as it rather is an CI issue than a build issue if the variable is not set correctly at all.

marbre replied:
With the above, `amdgpu_family` defaults to "unknown" and thus `not amdgpu_family` will never catch that is unset.

erman-gurses replied:
Forgot that part - updated

marbre commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -110,13 +115,15 @@ jobs:
 
       # TODO: Move to script
       - name: Create Index Files
+        if: always()
```
As you're assuming that the `logs` directory might not exist, it might be worth to rather guard this case here as the line `python build/indexer.py -f '*.log' build/logs/` as well as the call to sed wont work then.

erman-gurses replied:
Sounds good - handled that

marbre commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,49 @@
+    s3_base_path = f"s3://{bucket_name}/{run_id}/logs/{amdgpu_family}"
+
+    if not os.path.isdir(log_dir):
+        print("[INFO] No logs directory found. Creating it.")
```
Needs to be updated to reflect the latest changes. Furthermore, instead of using `print` you could consider introducing

```python
def log(*args, **kwargs):
    print(*args, **kwargs)
    sys.stdout.flush()
```

and use that like we do in other Python scripts.

erman-gurses replied:
using log now

marbre commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+import os
```
Can you add a brief description that include some usage instructions?

erman-gurses replied:
Added.

marbre commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -136,15 +149,23 @@ jobs:
         run: |
-          echo '* [Artifacts](https://therock-artifacts.s3.us-east-2.amazonaws.com/${{github.run_id}}/index-${{env.AMDGPU_FAMILIES}}.html)' >> $GITHUB_STEP_SUMMARY
-          echo '* [Build logs](https://therock-artifacts.s3.us-east-2.amazonaws.com/${{github.run_id}}/logs/${{env.AMDGPU_FAMILIES}}/index.html)' >> $GITHUB_STEP_SUMMARY
+          LOG_URL="https://s3.amazonaws.com/therock-artifacts/${GITHUB_RUN_ID}/logs/${AMDGPU_FAMILIES}/index.html"
```
The bucket URL is invalid:
> The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.

erman-gurses replied:
It came to my tired moment -  fixing it.

erman-gurses replied:
Fixed the URL

marbre commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -136,15 +149,23 @@ jobs:
+
+          ARTIFACT_INDEX=build/artifacts/index.html
+          if [ -f "${ARTIFACT_INDEX}" ]; then
+            ARTIFACT_URL="https://s3.amazonaws.com/therock-artifacts/${GITHUB_RUN_ID}/artifacts/${AMDGPU_FAMILIES}/index.html"
```
As above.

erman-gurses replied:
Fixed

erman-gurses replied:
Fixed the URL

marbre reviewed (changes requested):
There are some issues that need to be addressed before this can be merged. You might want to ask @ScottTodd for approval if you need another review today.

ScottTodd commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,74 @@
+    sys.stdout.flush()
+
+
+def run_aws_cp(src, dest, content_type=None):
```
When known, prefer to add type annotations to function arguments. Here it looks like the script is using `str` for `src` _and_ `dest`, but you could also use `Path` for `src`.

erman-gurses replied:
Good suggestion - added annotations.

ScottTodd commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,74 @@
+        dest: Destination S3 URI.
+        content_type: Optional MIME type to set in S3 metadata.
+    """
+    cmd = (
+        ["aws", "s3", "cp", src, dest, "--recursive"]
```
nit: could check that `aws` is available as a command to run using https://docs.python.org/3/library/shutil.html#shutil.which and provide a good error message if missing, or use the AWS Python SDK instead of the CLI.

erman-gurses replied:
Sounds good added a check using `shutil`

ScottTodd commented on .github/workflows/build_linux_packages.yml:
```diff
@@ -110,20 +115,36 @@ jobs:
+            echo "[INFO] Found 'build/logs' directory. Indexing '*.log' files..."
+            python build/indexer.py -f '*.log' build/logs/
+
+            if [ -f build/logs/index.html ]; then
+              echo "[INFO] Successfully created 'build/logs/index.html'. Updating links with AMDGPU_FAMILIES=${{ env.AMDGPU_FAMILIES }}..."
+              sed -i "s,a href=\"..\",a href=\"../../index-${{ env.AMDGPU_FAMILIES }}.html\",g" build/logs/index.html
+              echo "[INFO] Log index links updated."
+            else
+              echo "[WARN] 'build/logs/index.html' not found after indexing. Skipping link rewrite."
+            fi
```
+1 to simplifying here, but even better would be to move all of this logic into a Python script. Inlined bash within workflow files can't be easily iterated on or unit tested.

erman-gurses replied:
I simplified the script and testing it now. Will convert it to Python script then will do another test.

erman-gurses replied:
Converted to Python -  testing it.

erman-gurses commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,74 @@
+        log(f"[ERROR] Failed to upload {src} to {dest}: {e}")
+
+
+def upload_logs_to_s3(bucket_name, run_id, amdgpu_family, log_dir="build/logs"):
```
Refactored -  testing it now.

erman-gurses replied:
Test worked with the refactoring: https://github.com/ROCm/TheRock/actions/runs/15031434216

erman-gurses commented:
The last test completed fine with proper log files:
https://github.com/ROCm/TheRock/actions/runs/15035331614

marbre commented:
> @marbre, currently, `Build Windows Packages` on the main branch does not complete successfully. https://github.com/ROCm/TheRock/actions/runs/15035119910 If you want, I can raise another PR for `Build Windows Packages` after it has full functionality. What do you think?

It does succeed on the main branch on push, see https://github.com/ROCm/TheRock/actions/runs/15031073217/job/42243505016. However, there is obviously a permission issue when manually dispatching the workflow.

ScottTodd commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,101 @@
+    bucket_name: str, run_id: str, amdgpu_family: str, build_dir: Path
+):
+    log_dir = build_dir / "logs"
+    s3_base_path = f"s3://{bucket_name}/{run_id}-linux/logs/{amdgpu_family}"
```
We need handling in this script for both `-linux` and `-windows`. Can either infer that from the logs themselves, the sys/platform, or some explicit argument.

My intuition would be to have the workflow do this:

```bash
python build_tools/upload_logs_to_s3.py \
  --build-dir=${BUILD_DIR} \
  --s3-base-path="s3://therock-artifacts/${{github.run_id}}-linux/logs/${{env.AMDGPU_FAMILIES}}/"
```

Then the script would not need to know anything about amdgpu_family or run_id. Less plumbing that way, and the scripts would be runnable locally without needing to think about github actions. The disadvantage to that approach though is that the path structure would be only determined in a workflow file... maybe that could be its own `get_s3_base_path_for_workflow.py` script?

erman-gurses replied:
```
python build_tools/upload_logs_to_s3.py \
  --build-dir=${BUILD_DIR} \
  --s3-base-path="s3://therock-artifacts/${{github.run_id}}-linux/logs/${{env.AMDGPU_FAMILIES}}/"
```
This will reduce plumbing a lot. I will implement to see how it goes.

erman-gurses replied:
Addressed

ScottTodd commented on build_tools/create_log_index.py:
```diff
@@ -0,0 +1,48 @@
+        log(f"[WARN] Log directory '{log_dir}' not found. Skipping indexing.")
+        return
+
+    if index_file.exists():
+        log(
+            f"[INFO] Rewriting links in '{index_file}' with AMDGPU_FAMILIES={amdgpu_family}..."
+        )
+        content = index_file.read_text()
+        updated = content.replace(
+            'a href=".."', f'a href="../../index-{amdgpu_family}.html"'
+        )
```
Relating to my other comment, why do we rewrite these index file names? Aren't they already in separate subdirectories for each configuration?

ScottTodd commented:
Aside: I'm seeing lots of formatting commits pushed individually. If you locally set up pre-commit (or the specific formatters manually), that will let you automatically format your commits before you push.

erman-gurses commented:
> Aside: I'm seeing lots of formatting commits pushed individually. If you locally set up pre-commit (or the specific formatters manually), that will let you automatically format your commits before you push.

Yes, I will figure out automatically call black formatting.  ` locally set up pre-commit` sounds more comprehensive -  I can also try that - thank for pointing out.

erman-gurses commented:
> > @marbre, currently, `Build Windows Packages` on the main branch does not complete successfully. https://github.com/ROCm/TheRock/actions/runs/15035119910 If you want, I can raise another PR for `Build Windows Packages` after it has full functionality. What do you think?
> 
> It does succeed on the main branch on push, see https://github.com/ROCm/TheRock/actions/runs/15031073217/job/42243505016. However, there is obviously a permission issue when manually dispatching the workflow. This should be fixed with #622 but shouldn't block to add Windows specific adjustments to this PR anyway.

#622 helps a lot -  making progress on windows side.

erman-gurses commented:
Addressed all of the comments and able to generate both Windows and Linux log files.
Windows => https://github.com/ROCm/TheRock/actions/runs/15059473368
Linux => https://github.com/ROCm/TheRock/actions/runs/15059468317
@marbre @ScottTodd let me know If I am missing something.

ScottTodd commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,90 @@
+        run_aws_cp(str(index_path), index_s3_dest, content_type="text/html")
+        log(f"[INFO] Uploaded {index_path} to {index_s3_dest}")
+    else:
+        log("[INFO] No index.html found. Skipping index upload.")
```
Might want to report `log_dir` here, so it is clear where `index.html` was expected / looked for.

erman-gurses replied:
Sounds good just updated

ScottTodd commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,90 @@
+    parser.add_argument(
+        "--build-dir",
+        type=Path,
+        default=Path("build"),
```
This should anchor itself on the repository root, so running

```bash
cd build_tools
python upload_logs_to_s3.py
```

Doesn't search for logs in `build_tools/build`.

erman-gurses replied:
Updated

ScottTodd reviewed (changes requested):
Workflow changes LGTM. Just some Python style comments now.

erman-gurses commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,90 @@
+# === Updated build_tools/upload_logs_to_s3.py ===
+#!/usr/bin/env python3
```
My mistake - I keep track the file version manually that helps to go back quickly.

erman-gurses commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def check_aws_cli_available():
+    if shutil.which("aws") is None:
```
Updated

erman-gurses commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,90 @@
+def run_aws_cp(src: str, dest: str, content_type: str = None):
+    cmd = (
+        ["aws", "s3", "cp", src, dest, "--recursive"]
+        if os.path.isdir(src)
```
It make sense easier to use as a path  -  helps for the logic as well

erman-gurses commented on build_tools/upload_logs_to_s3.py:
```diff
@@ -0,0 +1,90 @@
+        sys.exit(1)
+
+
+def run_aws_cp(src: str, dest: str, content_type: str = None):
```
Sounds good  - updated

ScottTodd commented:
When responding to review feedback, you can batch your commit pushes to avoid re-triggering the CI so many times. Each push starts several hours of build jobs. Pushes to cancel previous jobs, but CI time is not cheap.

erman-gurses commented:
> When responding to review feedback, you can batch your commit pushes to avoid re-triggering the CI so many times. Each push starts several hours of build jobs. Pushes to cancel previous jobs, but CI time is not cheap.
Sure thing I can do that.

erman-gurses commented:
Final test after the last changes:
Windows: https://github.com/ROCm/TheRock/actions/runs/15076434735

ScottTodd commented on build_tools/create_log_index.py:
```diff
@@ -0,0 +1,73 @@
+def index_log_files(build_dir: Path, amdgpu_family: str):
+    log_dir = build_dir / "logs"
+    index_file = log_dir / "index.html"
+    indexer_path = build_dir / "indexer.py"
```
This still relies on the `indexer.py` file being downloaded from https://raw.githubusercontent.com/joshbrunty/Indexer/6d8cbfd15d3853b482e6a49f2d875ded9188b721/indexer.py ahead of time, right? We should still fork that file and customize it instead of having CI jobs download from github.

Fine for now, to keep this PR small, but could add a TODO.

erman-gurses replied:
I was planning to do that actually - will add TODO for that.

erman-gurses commented on build_tools/create_log_index.py:
```diff
@@ -0,0 +1,73 @@
+    return platform.system().lower() == "windows"
+
+
+def normalize_path(p: Path) -> str:
+    return str(p).replace("\\", "/") if is_windows() else str(p)
```
Ohh -  did not know that.

marbre replied:
Instead of forking `indexer.py` I propose to rather implement a solution based on boto3 which can then be used in an AWS Lambda.
